explorer-api/src/ping/http.rs

```diff
@@ -80,32 +82,112 @@ async fn port_check(bond: &MixNodeBond) -> HashMap<u16, bool> {
+    match lookup_host(host) {
+        Err(e) => {
+            error!("Failed to parse host {}:{} - {}", host, port, e);
+            None
         }
-        Ok(Err(_stream_err)) => {
-            error!("{} ping failed {:}", addr, _stream_err);
-            // didn't timeout but couldn't open tcp stream
-            false
-        }
-        Err(_timeout) => {
-            // timed out
-            error!("{} timed out {:}", addr, _timeout);
-            false
+        Ok(ip_addrs) => {
+            // DNS resolution must yield at least one ip address
+            if ip_addrs.is_empty() {
+                error!(
+                    "DNS lookup for host {}:{} didn't yield any IP addresses",
+                    host, port
+                );
+                return None;
+            }
+
+            // the host string should hopefully parse and resolve into a valid socket address
+            let parsed_host = format!("{}:{}", host, port);
+            match parsed_host.to_socket_addrs() {
+                Ok(mut addrs) => addrs.next(),
+                Err(e) => {
+                    error!(
+                        "Failed to resolve {}:{} -> {}. Error: {}",
+                        host, port, parsed_host, e
+                    );
+                    None
+                }
+            }
         }
     }
 }
+
+async fn do_port_check(host: &str, port: u16) -> bool {
+    match resolve_host(host, port) {
+        Some(addr) => match tokio::time::timeout(
+            CONNECTION_TIMEOUT_SECONDS,
+            tokio::net::TcpStream::connect(addr),
+        )
+        .await
+        {
+            Ok(Ok(_stream)) => {
+                // didn't timeout and tcp stream is open
+                trace!("Successfully pinged {}", addr);
+                true
+            }
+            Ok(Err(_stream_err)) => {
+                error!("{} ping failed {:}", addr, _stream_err);
+                // didn't timeout but couldn't open tcp stream
+                false
+            }
+            Err(_timeout) => {
+                // timed out
+                error!("{} timed out {:}", addr, _timeout);
+                false
+            }
+        },
+        None => false,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn resolve_host_with_valid_ip_address_returns_some() {
+        assert!(resolve_host("8.8.8.8", 1234).is_some());
+        assert!(resolve_host("2001:4860:4860::8888", 1234).is_some());
+    }
+
+    #[test]
+    fn resolve_host_with_valid_hostname_returns_some() {
+        assert!(resolve_host("nymtech.net", 1234).is_some());
+    }
+
+    #[test]
+    fn resolve_host_with_malformed_ip_address_returns_some() {
+        // these have missing values filled in
+        assert!(resolve_host("1.2.3", 1234).is_some()); // 1.2.0.3
+        assert!(resolve_host("1.2", 1234).is_some()); // 1.0.0.2
+        assert!(resolve_host("1", 1234).is_some()); // 0.0.0.1
```

Comment:
Why not `assert_eq!` on the exact `Some` value?

Reply:
Because the return type is `Option<SocketAddr>` I'm only interested that it returns a value. These tests are for edge cases such as bad IP address formatting, invalid characters, etc. 

To test equailty on `SocketAddr`, I'd have to be reverse looking up something stable to ensure the test passed - for example `nymtech.net` could move around or resolve to multiple addresses in a random order. And as you can see from the Windows failures, testing the OS with real data is useful.